### ISSUE TITLE
Add description to keymaps

### ISSUE THREAD

**Issue by LamprosPitsillos** [qutebrowser/qutebrowser#6981]
![keymaps_W](https://user-images.githubusercontent.com/61395246/151247435-609e0312-8774-44ae-b50e-761f0562fcdd.png)

Instead of showing the command that its going to call , lets have an optional argument `desc=""` that replaces the command in the keymap window with it self if not empty.If not set the command its going to be shown as usual.
```python
config.bind(
    leader + "dv",
    "hint links spawn --verbose yt-dlp {hint-url} --embed-thumbnail -o ~/vids/%(title)s.%(ext)s"
,# desc="Download Video"
)
```
```
    ,dv              Download Video
vs
    ,dv              hint links spawn --verbose yt-dlp {hint-url} --embed-thumbnail -o ~/vids/%(title)s.%(ext)s"

```

**Comment by The-Compiler** [comment]
I feel like we've had this before and there should be an issue around for it, but I can't seem to find it, so maybe I'm mistaken. cc @rcorre

**Comment by LamprosPitsillos** [comment]
I tried searching before making a new issue but also didn't find anything.

**Comment by rcorre** [comment]
It sounds like something we've done for completion, but I don't remember anything for the keyhint dialog. It sounds like a nice idea though!

**Comment by LamprosPitsillos** [comment]
I am thinking of trying a PR to address this issue , could you give me some resources on the matter? 
I did attempt this again months ago but got lost in the `asserts` and trying to navigated the code (python makes reading large code bases a bit difficult ..)

**Comment by The-Compiler** [comment]
@LamprosPitsillos don't think there's much generic stuff I can say - feel free to ask concrete questions (which you already did)! There's also [a chat](https://github.com/qutebrowser/qutebrowser/blob/master/doc/help/index.asciidoc#getting-help) where people can help if you want something more real-time.

---

I think one main question to consider is how to store those descriptions. This is probably something we should get a good solution for **before starting to write code**.

My suggestion would be to not cram this into `bindings.commands`, but instead make it a separate setting. Right now, `bindings.commands` is relatively simple (`{mode: {key: command}}`), adding additional information there would make it rather hard to manage.

However, if we have a *separate* `bindings.descriptions` setting, I see a couple of questions coming up:

- How does that setting look like? Does it map `{mode: {key: description}}`, or does it just map `{command: description}` instead? The latter seems simpler, but especially with commands like `hint links spawn --verbose yt-dlp {hint-url} --embed-thumbnail -o ~/vids/%(title)s.%(ext)s`, I see a risk of things getting out of sync when changing a binding
- How do we handle things getting out of sync? How do we *keep* them in sync? Does removing a binding also remove the description? What if there are multiple bindings? etc. etc.

---

Finally, I'm wondering: **Does this really provide enough value** to be worth the complexity from the user's side? We already have the `aliases` setting, so it would be already be possible to simply set up an alias for `download-video` to `hint links spawn --verbose yt-dlp {hint-url} --embed-thumbnail -o ~/vids/%(title)s.%(ext)s` and then bind a key to `download-video` instead. Having "Download video" instead of "download-video" seems like a quite minimal gain for the complexity this kind of thing would introduce (from an user's perspective, not necessarily in the code), no?

@LamprosPitsillos @rcorre @toofar what do you think?

**Comment by toofar** [comment]
TL;DR: yeah it does sound problematic to implement. Have a look at The-Compiler's suggestion of moving complicated commands into aliases (or userscripts) and see if that works well enough for you as a feature.

The proposal of
```python
config.bind(
    leader + "dv",
    "hint links spawn --verbose yt-dlp {hint-url} --embed-thumbnail -o ~/vids/%(title)s.%(ext)s",
    desc="Download Video",
)
```

Doesn't really fit into the current data model for that settings. Underneath it is stored as something like:

```python
bindings.commands = {
    "normal": {
        ";dv": "hint links spawn --verbose yt-dlp {hint-url} --embed-thumbnail -o ~/vids/%(title)s.%(ext)s",
     }
}
```

So to add a description into that you would have to add another layer of structure to the command definition (turn it into a tuple or another dict probably).
That would make it more cumbersome to set that setting using the way of settings commands using primitives like dict and list illustrated above. We could move that setting to be something you can only set via the python API to sidestep that problem, but that would be a change in API (and I don't like hiding settings behind the python API in general).
It's possibly that we could keep the settings-as-primitives view of things here but add a concept of extra data passed through the python config API that can be picked up by the resulting objects somehow. For example have some context object available where things like the [Command](https://github.com/qutebrowser/qutebrowser/blob/47ef2ca41d6f1f3981b4667105f9a1ce535beee6/qutebrowser/config/configtypes.py#L1036) object could opportunistically pull extra values it cared about from the context object. But that sounds like a whole project in itself to make that happen.

Adding a shadow `bindings.descriptions` setting could be made to work. And it might work fine for some users like you who care for this feature and put effort into maintaining your config around it. But it could also probably be error prone and confusing and we would have to answer questions about it and help users debug their config etc. Possibly adding some validation like "you have a description set for a binding that doesn't exist" would cover a fair amount of that. But it does seem a difficult API.

If you are looking at this as a way to contribute then feel free to keep discussing implementation ideas but unless a lot more people show interest or a straightforward implementation appears I wouldn't expect anything to get merged any time soon.

**Comment by LamprosPitsillos** [comment]
You are absolutely right , this isn't anything that important in the grand scheme of the project and i don't expect anyone from the core team to "waste" time on this, and aliases do sound like a good alternative for now.
> another example of why i would like to implement this feature,
![_W](https://user-images.githubusercontent.com/61395246/190625937-eafc13d4-e1c4-4cad-a97b-98a1476ea377.png)

But I am willing to mess around and try to contribute this functionality but my initial idea was this also
> So to add a description into that you would have to add another layer of structure to the command definition (turn it into a tuple or another dict probably).

And honestly I cant think of a better way right now but i will do some research on if there is a better way. Seeing as it is not that desirable of a solution by the maintainers.
The second solution 
> Adding a shadow `bindings.descriptions` setting could be made to work.

Sounds more cumbersome to me than just modifying (or potentially changing) the structure in which the keymaps are set.
As you have already mentioned, it will need a lot of maintenance from the developers in terms of inserting and deleting the descriptions accordingly ,and from the users to set the descriptions and unset them.

The only middle ground I can think of right now is a `Dict{mode: {key: description}}` in a `bindings.descriptions` that would simply check if the specific mode+key desc exist to print it on the key hint ,if it doesn't ,just print the command (In this way none of the logic of the existing binds has to change as the`bindings.descriptions` would be mostly used just for this cosmetic changes). The `DescDict` could be set in the `bind` functions as to keep the insertion simple :

> 
```python
config.bind(
    leader + "dv",
    "hint links spawn --verbose yt-dlp {hint-url} --embed-thumbnail -o ~/vids/%(title)s.%(ext)s"
,desc="Download Video"
)
```
And in the `unbind` function just search for the mode+key in the dict and if it exists remove it.

How does this sound? It will function on top of the existing code-base without needing any changes to the existing source code.

**Comment by The-Compiler** [comment]
Note that people don't necessarily will use `config.bind` - they might modify `bindings.commands` by hand, or use things like `:bind` in the UI. Any kind of "shadow structure" is almost bound to get out of sync with the changes to the actual keys of some sorts....

I still would much prefer to solve this via aliases somehow - i.e. something basically defining a new *command* (perhaps even with a description), not attaching the description to the keybindings.

So, suppose we had a `bindings.description_aliases` or so, which looks something like:

```
{
    "toggle-js-host": ("config-cycle -p -u *://{url:host}/* content.javascript.enabled ;; reload", "Toggle JS for the current host and save"),
    "toggle-js-host-temp": ("config-cycle -p -u *://{url:host}/* content.javascript.enabled ;; reload", "Toggle JS for the current host temporarily"),
    ....
}
```

those commands aren't shown in the `:` completion for the UI, but they will be accessible via keybindings - and the keyhint window would then show those descriptions instead of the command behind it.

Or, if we want to simplify it even more, we'd drop the description, and just show the "aliased" command. The main difference to "normal" aliases then being that the command isn't shown in the completions, to avoid cluttering it. At that point, we might as well just rename `aliases` to `aliases.commands`, and add this as `aliases.hidden` or somesuch.

Also, we'd then of course add such aliases for the more cumbersome default bindings.

**Comment by kuryart** [comment]
Sorry if I'm necrobumping, but what's the state of this? Can we define a description to keybindings via aliases? I still don't realize how.

**Comment by The-Compiler** [comment]
No, my description above was about a possible way of implementing this.

**Comment by Earnestly** [comment]
I do want to add a counter to this and ask for it not to be added (or at least can be disabled but this will add yet more test cases). My opinion on this isn't extremely strong but I haven't seen these points made before.

The clarity of the command is vastly more useful than a generic description (something I think systemd now allows showing the service's  filename instead of its useless (for me) description) as it tells me precisely what the browser is going to do.

Not only this but it also provides clues that allows me to learn the system more incrementally, how I might create my own commands and what is possible.

It doesn't seem like the additional work to support this is worth it. The difficulty of now describing commands (it is the hardest problem in computer science for a reason, describing abstractions clearly is hard), just to provide an _aesthetic_.

**Comment by LamprosPitsillos** [comment]
> I do want to add a counter to this and ask for it not to be added (or at least can be disabled but this will add yet more test cases). My opinion on this isn't extremely strong but I haven't seen these points made before.
> 
> The clarity of the command is vastly more useful than a generic description (something I think systemd now allows showing the service's filename instead of its useless (for me) description during startup) as it tells me precisely what the browser is going to do.
> 
> Not only this but it also provides clues that allows me to learn the system more incrementally, how I might create my own commands and what is possible.
> 
> It doesn't seem like the additional work to support this is worth it. The difficulty of now describing commands (it is the hardest problem in computer science for a reason, describing abstractions clearly is hard), just to provide an _aesthetic_.

I would have to disagree on this one, for some very simple reasons:

When i started using qutebrowser i had always used the very useful configFull.py that can be easily generated, it has everything you will ever need and more.

After using qutebrowser for years it's just ugly to have the command on the right instead of a plain description.
Let the default maps not have any , ok, that yould be a lot of work ( although it would be greatly appreciated ) , but like my original example, user commands, that I the user have created, i would prefer to have just a small description instead of 90 character command.

Neovim has something similar with is map API 

If i map a key to leader+d+m in qute to download music i would set a description like so "[d]ownload [m]usic"
Much better than a long ytdlp command.

**Comment by Earnestly** [comment]
This is still only for an aesthetic, rather than technical reason. You can solve the issue of long commands by considating them in userscripts and calling that, descriptively named, script.

*For those wondering about the systemd feature its configuration is `StatusUnitFormat=` in the `[Manager]` section of `/etc/systemd/system.conf`. See `systemd-system.conf(5)`.*

**Comment by LamprosPitsillos** [comment]
Can't argue with that, it is mostly about aesthetics, but also UX too. Descriptive names are much more user friendly than long commands.

And creating tens of small one-line user scripts where the name has to describe what it does isn't a very good solution either.
